Endpoint-specific dataset prefix

jotelha opened the pull request:
* Suggestion for introducing endpoint-specific DATASET_PREFIX configuration keys as found within `dtool-ecs`, in referenct o https://github.com/jic-dtool/dtool-s3/issues/11.
* Suggestion for "sanitizing" configuration keys into less problematic environment variable name, as contemplated at  https://github.com/jic-dtool/dtool-s3/pull/10#issuecomment-907783341

Note that I did not run the tests yet. I will have to set up some s3 testing environment and come back then.

tjelvar-olsson commented:
@jotelha great spot noticing that there was a lack of functionality in dtool-s3 with regards to what is present in dtool-ecs. This came about by dtool-s3 originally only having the ability to talk to AWS S3 buckets, i.e. no support for endpoints.

Now that we have migrated the support for alternative endpoints to dtool-s3 we should do as you suggest and add support for dataset prefixes on a per endpoint basis.

Looking at your code I don't understand the rational for introducing all the sanitization. When I tested this earlier I found that it worked out of the box. See: https://github.com/jic-dtool/dtool-s3/pull/10#issuecomment-907849003

I would argue that the only "code" line that needs changing is line 208 (from 8dc5ec0 diff) from:

`self.dataset_prefix = get_config_value("DTOOL_S3_DATASET_PREFIX")`

to:

`self.dataset_prefix = get_config_value("DTOOL_S3_DATASET_PREFIX_{}".format(bucket_name))`

@jotelha did you find a case where this was not good enough?

pastewka commented:
Hi @jotelha - can you address @tjelvar-olsson comments so we can get this into a release?

jotelha commented:
Sorry for the long delay here. There is one quite particular case for which I put that sanitation in the code. If for some reason (i.e. containerized environment), some endpoint-specific variables are specified both in `dtool.json` and the environment, then the environment variables should override the static config file content. On Unix-like systems, it's apparently possible to have environment variables that contain hyphens, but not that convenient, as shells don't support that. That means a simple

```console
$ export DTOOL_S3_ACCESS_KEY_ID_my-bucket=key
bash: export: `DTOOL_S3_ACCESS_KEY_ID_my-bucket=key': not a valid identifier
```

won't work, instead you need some detour like,

```console
$ env 'DTOOL_S3_ACCESS_KEY_ID_my-bucket=key' printenv | grep DTOOL
DTOOL_S3_ACCESS_KEY_ID_my-bucket=key
```

and I don't know how that situation looks like on other operating systems. In general, I like to stick to upper case letters and underscores for environment variables.

Hence, when I looked at @tjelvar-olsson's comment at https://github.com/jic-dtool/dtool-s3/pull/10#issuecomment-90778334 and the bucket naming rules at https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html, I thought that apart from the dot (.), bucket names specified via env variables can just as well use the conventional upper case letters (or any mix of upper and lower case letters) and underscores (_) instead of lower-case letters and hyphens (-) and still be mapped unambiguously onto proper bucket names (only containing lower case letters and hyphens)in the config files (and possibly override those settings).

Thus the sanitizing of the bucket name in the code. I am fine with getting rid of it for the merge and release. For me, that commit and comments here keep track of the subtle issue sufficiently :)

tjelvar-olsson commented:
@jotelha thank you for the explanation. I think what this highlights is that we need a better way to deal with these variables that are "bucket" specific. It has always been a bit ugly sticking the bucket name in the environment variable key. What is needed is something nested, but I don't know of a way to do that with environment variables.

If you could remove your sanitization for now. We can keep it in mind and try to come up with a better way of managing this information.

jotelha commented:
@tjelvar-olsson, removed the sanitization, note that I modified the README.rst, as the examples of type `export DTOOL_S3_ENDPOINT_my-bucket=...` won't work with a hyphen in the variable name.
